### PR TITLE
Create login system

### PR DESCRIPTION
**Pull request by radiantdinosaurs**

### CODE REVIEW

File: ui/app.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+})
+
+// error handler
+app.use(function(error, request, response) {
+    response.locals.message = error.message
+    response.locals.error = request.app.get('env') === 'development' ? error : {}
+    // render the error page
+    response.status(error.status || 500)
+    response.render('error')
+})
```
Comment (andrewvora) [inline ui/app.js]: This error handler should probably obfuscate all error information based on the environment. If it's `'development'` then it should give all info possible, but otherwise, the error message and stack trace should be hidden with some generic 404 page.

Also, doesn't seem like any of this gets called at all. At least, if you place a `console.log` sometime before `response.render('error')` and hit a nonexistent route like `/taco`. I would double check to see how to hand off the error from your middleware on L53

File: ui/app.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+app.use(session({
+    secret: '',
+    resave: true,
+    saveUninitialized: false,
+    cookie: {
+        maxAge: 3600000 // expires in one hour
+    },
+    store: new MongoStore({
+        url: dbUrl
+    })
```
Comment (andrewvora) [inline ui/app.js:35]: This is a suggestion for a different PR, but all configuration information like secrets and DB urls + credentials should be held in a config file or pulled from environment variables. If you keep it in a file, make sure that file is ignored and setup is detailed in a README local to the app.

File: ui/bin/www [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+server.listen(port)
+server.on('error', onError)
+server.on('listening', onListening)
+
+// normalize a port into a number, string, or false
+function normalizePort(val) {
+    const port = parseInt(val, 10)
+    if (isNaN(port)) return val
+    if (port >= 0) return port
+    return false
```
Comment (andrewvora) [inline ui/bin/www]: Rather than return false, maybe just thrown an error. If the port isn't set up, it's gonna thrown an error down the line anyways.

File: ui/controllers/userController.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
+'use strict'
```
Comment (andrewvora) [inline ui/controllers/userController.js]: Apparently, `use strict` is no longer necessary. So, I've read anyways. I have to remove it from my own projects, too

File: ui/controllers/userController.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+exports.logout = function logOut(request, response, next) {
+    if (request.session) {
+        request.session.destroy(function(error) {
+            if (error) {
+                logger.log('error', error)
+                return next(error)
+            } else return response.clearCookie('connect.sid', { path: '/' }).status(200).redirect('/')
+        })
+    } else response.status(200).redirect('/login')
+}
```
Comment (andrewvora) [inline ui/controllers/userController.js]: Filename should fall in line with the rest of the project.

`userController.js` -> `user-controller.js`

File: ui/controllers/userController.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                } else {
+                    // save the session for two weeks if the user checked the 'remember' box
+                    if (request.body.remember) request.session.cookie.maxAge = 14 * 24 * 3600000
+                    request.session.userId = user._id
+                    response.status(200).redirect('/profile')
+                }
+            })
+        }
+    }
+]
```
Comment (andrewvora) [inline ui/controllers/userController.js]: The final handler could be broken down into smaller functions. I would recommend breaking out the logic into another controller module. This should also make testing easier, instead of trying to wrestle out specific logic from this module

File: ui/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                } else {
+                    // save the session for two weeks if the user checked the 'remember' box
+                    if (request.body.remember) request.session.cookie.maxAge = 14 * 24 * 3600000
+                    request.session.userId = user._id
+                    response.redirect('/profile')
+                }
+            })
+        }
+    }
+]
```
Comment (andrewvora) [inline ui/controllers/userController.js]: Same comment about this handler, break out logic into separate controller and include tests

File: ui/controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        } else {
+            const userData = {
+                username: request.body.username,
+                display_name: request.body.username,
+                email: request.body.email,
+                password: request.body.password
+            }
+            User.create(userData, (error, user) => {
+                if (error) {
+                    if (error.code === 11000) {
```
Comment (andrewvora) [inline ui/controllers/userController.js]: If this number was pulled from documentation somewhere, it would be better to define a top-level `const` at the top of this module and in the comments link the documentation that you pulled this value from

File: ui/models/user.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
+'use strict'
```
Comment (andrewvora) [inline ui/models/user.js]: Same note about `use strict` as earlier

File: ui/models/user.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                    error.status = 401
+                    callback(error)
+                }
+            })
+        }
+    })
+}
+
+const User = mongoose.model('Users', UserSchema)
+module.exports = User
```
Comment (andrewvora) [inline ui/models/user.js]: Since there's validation logic in here, you should move the functions into their own module (like `mongoose-user-validation.js`) that exports the functions so you can use them in this `User` module. This way, you can test that logic without messing with Mongoose's magic

Comment (andrewvora) [reply ui/models/user.js]: Also, it'll come with the added benefit of moving some `require` dependencies out of this file and into the validation module

**Review by andrewvora — CHANGES_REQUESTED** [review]
The logic itself seems sound, but there's lot of logic and no tests.

The bigger issue is that this kind of subverts the project direction by including DB and authentication logic into the UI app. The intent was to have all of that into the app module as additional routes and have UI only handle the display of data from the server app. Session management should be handled in the UI, but instead of accessing the database, you'll be hitting the server's API to get things like user objects and security tokens.

The "frontend" UI node app can do database operations, but nothing that's imperative to the slothbucket server. Keep it limited to UI-related operations.

**tl;dr** move all DB and auth logic into the `app/` dir and add tests. Continue managing session in the `ui/` dir.

File: app/controllers/auth-controller.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            response.status(400).send({status: 400, error: errors.array()})
+        } else {
+            const userData = {
+                username: request.body.username,
+                display_name: request.body.username,
+                email: request.body.email,
+                password: request.body.password
+            }
+            userController.createUser(userData).then((user) => {
+                console.log(process.env.SECRET)
```
Comment (andrewvora) [inline app/controllers/auth-controller.js]: Instead of a just printing it out (which is a security risk) shouldn't you instead have handling logic? Something like `process.exit()` if it's not set? Since without it, your auth system shouldn't be functioning

File: app/app.js [hunk 12]
```diff
@@ -2,74 +2,40 @@
...
-                    .send({status: error.code, message: 'Internal error encountered. Please try again.'})
-            })
-        }).catch(error => {
-            if (error.code === undefined) error.code = 500
-            if (error.code === 500) logger.log('error', error)
-            else logger.log('warn', error)
+const router = require('./routes/router')
+app.use('/', router)
+app.use((request, response, next) => next(returnError.resourceNotFound()))
+app.use((error, request, response, next) => {
```
Comment (andrewvora) [inline app/app.js]: These two handlers could use comments, so others wouldn't need to look up what they're used for

File: ui/app.js [hunk 13]
```diff
@@ -36,33 +34,29 @@ app.use(session({
...
-app.use(function(error, request, response) {
-    response.locals.message = error.message
-    response.locals.error = request.app.get('env') === 'development' ? error : {}
-    // render the error page
-    response.status(error.status || 500)
-    response.render('error')
+const router = require('./routes/router')
+app.use('/', router)
+app.use((request, response, next) => next(returnError.resourceNotFound()))
+app.use((error, request, response, next) => {
```
Comment (andrewvora) [inline ui/app.js]: These two handlers could use comments describing their function

File: ui/controllers/auth-controller.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                    } else if (body.auth && body.token && body.user_id) {
+                        request.session.userId = body.user_id
+                        request.session.jwt = body.token
+                        response.status(200).redirect('/')
+                    } else next(returnError.backendError())
+                }
+            })
+        }
+    }
+]
```
Comment (andrewvora) [inline ui/controllers/auth-controller.js]: The logic here could be broken out into smaller functions. Pass variables you need as params. It's difficult to read the logic without having to find out everything `post_register` does

If it makes it easier for you, separate the exports with multiline comments like

```javascript
/*------------------------*
 * POST /login
 *-------------------------*/
```

So the functions don't become hard to group

File: ui/controllers/auth-controller.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+                    if (body.error) {
+                        // specific error message from the server, usually pertaining to incorrect username/password
+                        error = [{msg: body.error}]
+                        response.status(200).render('login', {title: 'Login', errors: error})
+                    } else if (body.auth && body.token && body.user_id) {
+                        request.session.userId = body.user_id
+                        request.session.jwt = body.token
+                        console.log(request.session)
+                        response.status(200).redirect('/')
+                    } else next(returnError.backendError())
```
Comment (andrewvora) [inline ui/controllers/auth-controller.js]: Same here. Since the callback and if-elses already require so much nesting, any way you can flatten the structure would make it easier to read

File: app/controllers/user-controller.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+            } else if (!user) {
+                const error = returnError.incorrectUsernameOrPassword()
+                reject(error)
+            } else resolve(user)
+        })
+    })
+}
+
+module.exports.createUser = createUser
+module.exports.findUser = findUser
```
Comment (andrewvora) [inline app/controllers/user-controller.js]: `user-controller` could use some tests

File: ui/controllers/auth-controller.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+exports.logout = function logOut(request, response, next) {
+    if (request.session) {
+        request.session.destroy(function(error) {
+            if (error) {
+                logger.log('error', error)
+                return next(error)
+            } else return response.clearCookie('connect.sid', { path: '/' }).status(200).redirect('/')
+        })
+    } else response.status(200).redirect('/login')
+}
```
Comment (andrewvora) [inline ui/controllers/auth-controller.js]: `auth-controller` could use some tests

**Review by andrewvora — CHANGES_REQUESTED** [review]
It's looking a lot better, but still has a lot of holes in terms of tests. There's a lot of code being added and, in the future, you want to be sure that upcoming changes aren't breaking anything vital. A broken test serves the purpose of telling a developer (hey, you broke something) and making them ask themselves, "did I mean to break this?"

**Review by andrewvora — APPROVED** [review]
Let's get this merged in
